### PR TITLE
fix: mandatory date column enforcement

### LINKED ISSUE #34258: [Bug]: Date column set as required does not making it mandatory for adding new Table row
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

A user has a Table widget with a date column, and has enabled `Adding a row`. He has set the date column to be required.
But when adding a new row, and not giving any value to the Date column, he is still able to click on Save row button. And there is no error shown stating that the Date field is required.

### Steps To Reproduce

1. Drag and drop a table widget
2. Connect data to it and enable Add new Row
3. Configure one of the columns as Date column and set it as required
4. Try adding a new Row without setting value for the date column

You can see, the `Save row` is not disabled, even though I haven't given any value for the Date column which is set as required.

**Note**: This issue is not seen while editing a Table row as the error message is seen when no value is given.

### Public Sample App

_No response_

### Environment

Production

### Severity

Medium (Frustrating UX)

### Issue video log

_No response_

### Version

Cloud

### PR DESCRIPTION
**Pull request by jacquesikot**
## Description
**Problem:**
When the `isRequired` property is set for date columns in the table widget, the validation doesn't work as expected. Users can add new rows without filling in the date field, even though it is marked as required. This results in rows being added with missing date values, which can lead to incomplete or invalid data entries.

**Root Cause:**
The validation logic for the date column is currently handled within the DateCell component. However, the isRequired validation functionality was not implemented within this component. Additionally, the general validation logic in the `getEditableCellValidity` function, located in the derived.js file, does not account date cells in its isRequired validation.

**Solution:**
To fix this issue:

**Enhance the `getEditableCellValidity` function:** Extend the existing validation logic in getEditableCellValidity to include the date cell in its validation, specifically checking for the isRequired property.

**Integrate with DateCell validation:** Ensure that the isRequired validation is properly executed in conjunction with the existing date validations inside the DateCell component. This will enforce the requirement and prevent new rows from being added if the date field is left empty.

**Test Plan**
https://www.notion.so/appsmith/Test-Plan-Date-Column-Marked-as-Required-Doesn-t-Enforce-Mandatory-Entry-When-Adding-New-Table-Row-c73b764af60842a188cba056bdda6d79?pvs=4

Fixes #34258

## Automation

/ok-to-test tags="@tag.All"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/10453174231>
> Commit: 40fe2eaf7d45024bef00f8031e971a5bac2a4e3e
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=10453174231&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.All`
> Spec:
> <hr>Mon, 19 Aug 2024 13:17:23 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Summary by CodeRabbit

- **New Features**
  - Introduced validation for "Date" column types, ensuring that required fields are correctly enforced.

- **Enhancements**
  - Improved validation logic for more accurate user feedback in date cells.
  - Expanded support for validating "date" columns in the table widget.

- **Bug Fixes**
  - Enhanced error handling to ensure proper indication of cell validity based on new validation criteria.
  - Updated visual feedback for cell editor state in the UI.

### CODE REVIEW

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/AddNewRow2_spec.js [hunk 1]
```diff
@@ -87,6 +87,18 @@ describe("Validation flow", { tags: ["@tag.Widget", "@tag.Table"] }, () => {
...
+    cy.wait("@updateLayout");
+    cy.enterTableCellValue(0, 0, "");
+    cy.get(`.t--inlined-cell-editor-has-error`).should("exist");
+
+    // revert to Number for remainder of tests
+    cy.get(commonlocators.changeColType).last().click();
+    cy.get(".t--dropdown-option").children().contains("Number").click();
+    cy.wait("@updateLayout");
+
     cy.get(".t--discard-new-row").click({ force: true });
```
Comment (sagar-qa007) [inline app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/AddNewRow2_spec.js:102]: @jacquesikot please add validation after this action. And also update the test title if needed.

**Review by sagar-qa007 — CHANGES_REQUESTED** [review]

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/AddNewRow2_spec.js (hunk 1 above)
Comment (jacquesikot) [reply app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/AddNewRow2_spec.js:102]: @sagar-qa007 This is to revert to the Number state that is used for the remainder of the test cases below. I am not asserting or validation anything with the statements, just returning test to original state that the remaining cases need.

**Review by sagar-qa007 — DISMISSED** [review]

**Review by rahulbarwal — DISMISSED** [review]
Only remaining thing: a feature flag. Otherwise its good.

**Review by rahulbarwal — DISMISSED** [review]

**Review by rahulbarwal — APPROVED** [review]

**Pull request merged**